ethers-providers/src/transports/retry.rs

```diff
@@ -0,0 +1,257 @@
+    fn next_backoff(&mut self, total_enqueued_requests: u32, retry_number: u64)
+        -> Option<Duration>;
+}
+
+/// [RetryClient] presents as a wrapper around [JsonRpcClient] that will retry
+/// requests based on [Backoff] strategy and filtering based on [RetryPolicy].
+#[derive(Debug)]
+pub struct RetryClient<T, U, B>
+where
+    T: JsonRpcClient,
+    T::Error: Sync + Send + 'static,
+    U: RetryPolicy<T::Error> + Send + Sync + Debug,
+    B: Backoff + Send + Sync,
+{
+    inner: T,
+    requests_enqueued: AtomicU32,
+    policy: U,
+    backoff: Mutex<B>,
+}
+
+impl<T, U, B> RetryClient<T, U, B>
+where
+    T: JsonRpcClient,
+    T::Error: Sync + Send + 'static,
+    U: RetryPolicy<T::Error> + Send + Sync + Debug,
+    B: Backoff + Send + Sync,
+{
+    /// Example:
+    ///
+    /// ```no_run
+    /// let http = Http::new(Url::parse("http://localhost:8545").unwrap());
+    /// let delay = Duration::new(10, 0);
+    /// let client = RetryClient::new(http, HttpRateLimitRetryPolicy::new(), DelayedQueuedBackoff::new(delay));
+    /// ```
+    pub fn new(inner: T, policy: U, backoff: B) -> Self {
+        Self { inner, requests_enqueued: AtomicU32::new(0), policy, backoff: Mutex::new(backoff) }
+    }
+}
+
+/// Error thrown when:
+/// 1. Internal client throws an error we do not wish to try to recover from.
+/// 2. Params serialization failed.
+/// 3. Request timed out i.e. `Backoff::next_backoff` returned `None`.
+#[derive(Error, Debug)]
+pub enum RetryClientError<T>
+where
+    T: JsonRpcClient,
+    T::Error: Sync + Send + 'static,
+{
+    #[error(transparent)]
+    ProviderError(T::Error),
+    TimeoutError,
+    #[error(transparent)]
+    SerdeJson(serde_json::Error),
+}
+
+impl<T> std::fmt::Display for RetryClientError<T>
+where
+    T: JsonRpcClient,
+    <T as JsonRpcClient>::Error: Sync + Send + 'static,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl<T> From<RetryClientError<T>> for ProviderError
+where
+    T: JsonRpcClient + 'static,
+    <T as JsonRpcClient>::Error: Sync + Send + 'static,
+{
+    fn from(src: RetryClientError<T>) -> Self {
+        ProviderError::JsonRpcClientError(Box::new(src))
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<T, U, B> JsonRpcClient for RetryClient<T, U, B>
+where
+    T: JsonRpcClient + 'static,
+    T::Error: Sync + Send + 'static,
+    U: RetryPolicy<T::Error> + Send + Sync + Debug,
+    B: Backoff + Send + Sync + Debug + Clone,
+{
+    type Error = RetryClientError<T>;
+
+    async fn request<A, R>(&self, method: &str, params: A) -> Result<R, Self::Error>
+    where
+        A: std::fmt::Debug + Serialize + Send + Sync,
+        R: DeserializeOwned,
+    {
+        self.requests_enqueued.fetch_add(1, Ordering::SeqCst);
+
+        let params =
+            serde_json::to_value(params).map_err(|err| RetryClientError::SerdeJson(err))?;
```

Comment:
this works,
but now that I see it, I believe serialize to String would be more efficient because then we don't have to serialize Value again and again